### DISCUSSION TITLE
[SuperiorSkyblockAPI] New events and methods

### DISCUSSION THREAD

**Discussion by Ceduz19** [BG-Software-LLC/SuperiorSkyblock2#1825] [Pending Feature Requests] [answered]
I am currently developing a small addon for SuperiorSkyblock and while doing so I found various things that I needed missing in the API.

## Events
Currently the API has only events related to when the plugin is initializing or initialized (_PluginInitializeEvent_ and _PluginInitializedEvent_) and when the data manager is loaded or reloaded (_PluginLoadDataEvent_).
There are no events called when the plugin is **reloaded** by using its command (_/is admin reload_) so I'm requesting to add two simple events:

### PluginReloadEvent
An event called before actually reloading, similarly to PluginInitializeEvent; it may be useful also to have the possibility to change providers or other settings when called.

### PluginReloadedEvent
As you can deduce, this is an event called when SuperiorSkyblock finished reloading all its stuff.

## Methods
While developing my addon I thought of some simple (maybe) methods that can be really useful in some situations.
_The names of the following methods are the first ones that came to my mind so i know they can be really bad but the point is to give the idea._

### getDatabaseSession() method
As the name says, this method will return the current running SQL session, this would permit to devs to manipolate as they wish data in the same database used by SuperiorSkyblock being able to also create new tables with island-related data.

I know that currently you can get from the API the database settings in the config permitting you to establish by yourself a connection with it but this would need some extra work that could be easily avoided by introducing this method.

### getMessage(Message, Locale) method
This method will allow you to obtain the string, or the list, set for the _message_ specified (one of the Message's enums) in one of the lang files depending on the _locale_ argument.

### setMessage(Message, Locale, MessageComponent) method
In this case using this method you will have the possibility to set to the _message_ for the _locale_ specified a _message component_ built by you (this would maybe need the following method, otherwise instead of the MessageComponent you'll need a string or string list).

### getMessagesService() method
Just for utility, the method will return the MessagesService used by the plugin in order to easily get or parse components (this would mean also to add the various MessageComponents used by SuperiorSkyblock).
_Not really needed method but always useful to have._

These are my API-related features requests, I'm aware that some of them could be considered useless or not really necessary but I wanted to write them anyway.
Thanks for reading, Ceduz19.

**Comment by OmerBenGera** [comment, marked as answer]
Hey, thanks for all the suggestions!

### Reload Events

If you develop a module for the plugin, you already have such thing. If you don't use the modules system, then you should if it is an addon dedicated for SSB only.

### getDatabaseSession() method

The database can be other things than SQL, therefore such thing will never be exposed to the API.
You can use the database bridges to manipulate the DB.
Btw, you should not do such thing and instead use the persistent data API to track data regarding players / islands.

### getMessage(Message, Locale) method

Already exists - https://bg-software.com/superiorskyblock/api/
Look for MessageService#getComponent

### setMessage(Message, Locale, MessageComponent) method

Messages are by-design only editable by the player, through the config files.
You can however have your own messages that can be parsed easily using the API similar to SSB's syntax. Look at MessageService.

### getMessagesService() method

This already exists. Look at Bukkit's services system.
```Bukkit.getServiceManager().getRegistration(MessageService.class)```

**Reply by Ceduz19** [reply]
Thank you so much for responding and I apologize for suggesting things that already exist (I really tried to find them in the API before writing this but I didn't succeed; I'm probably blind) or are not really convenient to implement.
Thanks again, Ceduz19.
